\documentclass{article}[12pt]


\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage[english,french]{babel}
\usepackage{url}
\usepackage{scrextend}
\addtokomafont{labelinglabel}{\sffamily}

\begin{document}

<<setup, include=FALSE, cache=FALSE>>=
library(knitr)
# set global chunk options
opts_chunk$set(fig.path='figure/minimal-', fig.align='center', fig.show='hold')
options(formatR.arrow=TRUE,width=90)
@


\title{An Introduction to PatPilr}

\author{Rapha\"el Jauslin}

\maketitle

\newpage

\tableofcontents
\newpage

\section{Overview}

This package is a tool to facilitate the pre-treatment and the treatment of NGS data. The tools are implemented to work on fastq and fasta file. This introduction will, step-by-step, explain how to use the package and which functions you should use in order to obtain your data merged, demultiplexed and cleaned. This will supposed that you have two .fastq R1 and R2 and a barcode file that contains the informations for the demultiplexing step. Firstly, we need to download and install the package PatPilr. You could find the package in the github repository : \url{https://github.com/RJauslin/PatPilr}. You should launch the following commands in \texttt{R} or \texttt{Rstudio} in order to install PatPilr.

<<install,cache=FALSE,eval=FALSE>>=
install.packages("devtools")
devtools::install_github("Rjauslin/PatPilr@master")
@

\section{Pre-treatment}
\label{sec:pretreatment}
This is the first step of the pipeline. The input are the original files from the server they are generally named \texttt{xxx\_R1.fastq} and \texttt{xxx\_R2.fastq}. We will explain how the program deal with the main three steps, merging, demultiplexing, and cleaning. During all the process we will supposed that you have put only the two fastq files and the informations needed for the demultiplexing (see Section \ref{subsec:demux}).

<<load,cache=FALSE,eval=FALSE>>=
library(PatPilr)

#Linux
pathFolder <- "/home/raphael/Documents/......./working_directory/"
#Windows
pathFolderWindows <- "C:/Users/raphael/......./working_directory/"

@

All the pre-treatment is wrap inside a function that call the different functions. So you only have to check the parameter of the function and click enter.

<<mainfunc,cache=FALSE,eval=FALSE>>=
preTreatment(pathFolder,
  m = 10, # min overlap
  M = 100, # max overlap
  x = 0.25, # max mismatch density
  t = 4, # number of threads
  mismatch = FALSE, # allows 1 mismatch in tag if TRUE
  err = 0.01, #
  slide = 50, #
  minlength = 60) # parameters for qual check
@


\subsection{Merging}
\label{subsec:merge}

The merging step currently implemented is done by the program FLASH \cite{Magoc2011}. We have allows some possible parameters

\begin{labeling}{mcccc}
\item [m] The minimum required overlap length between two reads to provide a confident overlap.
\item [M] Maximum overlap length expected in approximately 90\% of read pairs.
\item [x] Maximum allowed ratio between the number of mismatched base pairs and the overlap length.
\item [t] Set the number of worker threads.
\end{labeling}

\subsection{Demultiplexing}
\label{subsec:demux}


\subsubsection{Simple tag}
The demultiplexing step is implemented by the program \texttt{PatPil} that is hide in the package. The function calls the tools \texttt{D\_simple\_tag} that could be used from the shell by the following command.

<<demultiplex_simple,cache=FALSE,eval=FALSE,size ='footnotesize'>>=
./PatPil D_simple_tag -f ./merged.fastq -o ./outputFolder/ -b ./barcodes.txt -mismatch
@

The only thing that you should care is that your barcode.txt file is of the following form.

<<barcode,cache=FALSE,eval=FALSE>>=
ACGAGTGCGT	01.fq
ACGCTCGACA	02.fq
AGACGCACTC	03.fq
AGCACTGTAG	04.fq
ATCAGACACG	05.fq
ATATCGCGAG	06.fq
CGTGTCTCTA	07.fq
CTCGCGTGTC	08.fq
...
@

\textbf{More important the separator between the tags and the names of the files should be a tab.}

\subsubsection{Double tag}


\subsection{Quality check}
\label{subsec:qualcheck}

\cite{Edgar2015}



\section{Dereplication}






\bibliographystyle{apalike} % Le style est mis entre accolades.
\bibliography{patpil} % mon fichier de base de donn√©es s'appelle bibli.bib

\end{document}
